python/message.c

```diff
@@ -0,0 +1,1487 @@
+  PyUpb_CMessage* self = (PyUpb_CMessage*)_self;
+  if (upb_fielddef_issubmsg(field) || upb_fielddef_isseq(field)) {
+    PyErr_Format(PyExc_AttributeError,
+                 "Assignment not allowed to message, map, or repeated "
+                 "field \"%s\" in protocol message object.",
+                 upb_fielddef_name(field));
+    return -1;
+  }
+
+  PyUpb_CMessage_AssureWritable(self);
+
+  upb_msgval val;
+  upb_arena* arena = PyUpb_Arena_Get(self->arena);
+  if (!PyUpb_PyToUpb(value, field, &val, arena)) {
+    return -1;
+  }
+
+  upb_msg_set(self->msg, field, val, arena);
+  return 0;
+}
+
+int PyUpb_CMessage_GetVersion(PyObject* _self) {
+  PyUpb_CMessage* self = (PyUpb_CMessage*)_self;
+  return self->version;
+}
+
+/*
+ * PyUpb_CMessage_GetAttr()
+ *
+ * Implements:
+ *   foo = msg.foo
+ *
+ * Attribute lookup must find both message fields and base class methods like
+ * msg.SerializeToString().
+ */
+__attribute__((flatten)) static PyObject* PyUpb_CMessage_GetAttr(
+    PyObject* _self, PyObject* attr) {
+  PyUpb_CMessage* self = (PyUpb_CMessage*)_self;
+
+  // Lookup field by name.
+  const upb_fielddef* field;
+  if (PyUpb_CMessage_LookupName(self, attr, &field, NULL, NULL)) {
+    return PyUpb_CMessage_GetFieldValue(_self, field);
+  }
+
+  // Check base class attributes.
+  assert(!PyErr_Occurred());
+  PyObject* ret = PyObject_GenericGetAttr(_self, attr);
+
+  // Return value if found, swallow AttributeError if raised.
+  if (ret) {
+    return ret;
+  }
+  if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
+    return NULL;
+  }
+  PyErr_Clear();
+
+  return PyUpb_MessageMeta_GetAttr((PyObject*)Py_TYPE(_self), attr);
+}
+
+/*
+ * PyUpb_CMessage_SetAttr()
+ *
+ * Implements:
+ *   msg.foo = foo
+ */
+static int PyUpb_CMessage_SetAttr(PyObject* _self, PyObject* attr,
+                                  PyObject* value) {
+  PyUpb_CMessage* self = (PyUpb_CMessage*)_self;
+  const upb_fielddef* field;
+  if (!PyUpb_CMessage_LookupName(self, attr, &field, NULL,
+                                 PyExc_AttributeError)) {
+    return -1;
+  }
+
+  return PyUpb_CMessage_SetFieldValue(_self, field, value);
+}
+
+static PyObject* PyUpb_CMessage_HasField(PyObject* _self, PyObject* arg) {
+  PyUpb_CMessage* self = (PyUpb_CMessage*)_self;
+  const upb_fielddef* field;
+  const upb_oneofdef* oneof;
+
+  if (!PyUpb_CMessage_LookupName(self, arg, &field, &oneof, PyExc_ValueError)) {
+    return NULL;
+  }
+
+  if (field && !upb_fielddef_haspresence(field)) {
+    PyErr_Format(PyExc_ValueError, "Field %s does not have presence.",
+                 upb_fielddef_fullname(field));
+    return NULL;
+  }
+
+  if (PyUpb_CMessage_IsUnset(self)) Py_RETURN_FALSE;
+
+  return PyBool_FromLong(field ? upb_msg_has(self->msg, field)
+                               : upb_msg_whichoneof(self->msg, oneof) != NULL);
+}
+
+static PyObject* PyUpb_CMessage_ListFields(PyObject* _self, PyObject* arg) {
+  PyObject* list = PyList_New(0);
```

Comment:
Can you push this into the branches so we can give a size estimate to PyList to start with?

Reply:
I think we have two cases
1) msg == null in which case our list should start empty
2) msg != null in which case we will have field_count items in the list

Is there some way we don't end up with one of those two?